staking/slash/double-sign.go

```diff
@@ -1,52 +1,301 @@
+		amt := numeric.NewDecFromBigInt(
+			delegator.Amount,
+		)
+		discounted := amt.Mul(rate).TruncateInt()
+		half := numeric.NewDecFromBigInt(
+			new(big.Int).Sub(delegator.Amount, discounted),
+		).Mul(fiftyPercent)
+
+		paidOff := numeric.ZeroDec()
+
+		// NOTE Get the token trying to escape
+		for _, undelegate := range delegator.Undelegations {
+			// the epoch matters, only those undelegation
+			// such that epoch>= doubleSignEpoch should be slashable
+			if undelegate.Epoch.Cmp(doubleSignEpoch) == -1 {
+				continue
+			}
+			if paidOff.Equal(half) {
+				break
+			}
+			undelegate.Amount.SetInt64(0)
+			slashTrack.TotalSlashed.Add(slashTrack.TotalSlashed, undelegate.Amount)
+			paidOff = paidOff.Add(numeric.NewDecFromBigInt(undelegate.Amount))
+		}
+
+		halfB := half.TruncateInt()
+		state.AddBalance(reporter, halfB)
+		slashTrack.TotalSnitchReward.Add(slashTrack.TotalSnitchReward, halfB)
+
+		// we might have had enough in undelegations to satisfy the slash debt
+		if paidOff.Equal(half) {
+			continue
+		}
+		// otherwise we need to dig deeper in their pocket
+		leftoverDebt := amt.Sub(half.Sub(paidOff)).TruncateInt()
+		delegator.Amount.Sub(delegator.Amount, leftoverDebt)
+		paidOff.Add(numeric.NewDecFromBigInt(leftoverDebt))
+		slashTrack.TotalSlashed.Add(slashTrack.TotalSlashed, leftoverDebt)
+
+		if paidOff.Equal(half) {
+			fmt.Println("special case when exactly equal", paidOff.String(), half.String())
+		}
+
+		if !paidOff.GTE(half) {
+			fmt.Println(leftoverDebt.String(), "\n Now have ",
+				delegator.Amount, "-had-",
+				amt.String(),
+				paidOff,
+				half,
+				delegations.String(),
+			)
+
+			return errors.Errorf(
+				"paidoff %s is not >= slash debt %s",
+				paidOff.String(),
+				half.String(),
+			)
+		}
+	}
+	fmt.Println("after delegator slash application", slashTrack.String())
 	return nil
 }
+
+// TODO Need to keep a record in off-chain db of all the slashes?
+
+// Apply ..
+func Apply(
+	chain staking.ValidatorSnapshotReader, state *state.DB,
+	slashes []Record, committee []shard.BlsPublicKey,
+) (*Application, error) {
+	log := utils.Logger()
+	rate := Rate(uint32(len(slashes)), uint32(len(committee)))
+	slashDiff := &Application{big.NewInt(0), big.NewInt(0)}
+
+	for _, slash := range slashes {
+		// TODO Probably won't happen but we probably should
+		// be expilict about reading the right epoch validator snapshot,
+		// because it needs to be the epoch of which the double sign
+		// occurred
+		snapshot, err := chain.ReadValidatorSnapshot(
+			slash.Offender,
+		)
+
+		if err != nil {
+			return nil, errors.Errorf(
+				"could not find validator %s",
+				common2.MustAddressToBech32(slash.Offender),
+			)
+		}
+
+		nowAccountState := state.GetStakingInfo(slash.Offender)
+
+		if nowAccountState == nil {
+			return nil, errors.New("cannot be nil code field in account state")
+		}
+
+		const mustBeValidatorsDelegationToSelf = 1
+		// Handle validator's delegation explicitly for sake of clarity
```

Comment:
don't need to handle validator's stake separately. Just do slashing on all Delegations[] together.